### PR TITLE
refactor(ERTP,vats): AmountPatternShape

### PR DESCRIPTION
**Pull request by erights**
closes: XXX
refs: https://github.com/Agoric/agoric-sdk/pull/9402#discussion_r1612477663 https://github.com/Agoric/agoric-sdk/pull/9402#discussion_r1612651947  https://github.com/Agoric/agoric-sdk/issues/9407

## Description

#9407 explains that sometimes we used a `.optional(AmountShape)` guard to describe an `optAmountShape` argument.

https://github.com/Agoric/agoric-sdk/pull/9402#discussion_r1612477663 https://github.com/Agoric/agoric-sdk/pull/9402#discussion_r1612651947 discussed ways to fix this that were ultimately omitted from #9402 . This PR provides a form of the fix discussed in those comments. Three remaining controversies I'd like my reviewers to comment on:
- Whether the name of the exported pattern should be `AmountShapeShape` or `AmountPatternShape`. See my doc-comment on `AmountPatternShape` for my weak reasoning about why I chose this name.
- If we do go with `AmountPatternShape` for the stated reasons, should the corresponding parameter variables be changed from `optAmountShape` to `optAmountPattern`.
- Whatever the name of the exported pattern, should it be defined as 
   - `M.pattern()`
   - `harden({ brand: M.pattern(), value: M.pattern() })`
   - `harden({ brand: BrandShape, value: M.pattern() })`
   - something else

### Security Considerations

For almost all of the choices of resolving the above controversies, this PR remains a pure bug fix. All correct code that used to work will continue working the same way, and some correct code that used to incorrectly fail due to this bug will now start working correctly.

The exception would be if we both accepting the renaming of some existing occurrences of `M.pattern()` to `AmountPatternShape`/`AmountShapeShape`, and define this name as a pattern narrower than `M.pattern()`. In that case, existing (hypothetical) code that, for example, used `M.any()` in such an argument position would start failing.

### Scaling Considerations

None
### Documentation Considerations

None
### Testing Considerations

#9407 mentions how to reproduce the bug it reports. This PR should add that as a test, to verify that this PR fixes that bug.
### Upgrade Considerations

https://github.com/Agoric/agoric-sdk/pull/9402#discussion_r1613503324 and https://github.com/Agoric/agoric-sdk/pull/9402#discussion_r1613622732 explain the upgrade concern that was likely the cause for omitting this fix from #9042

### CODE REVIEW

File: packages/internal/src/marshal.js [hunk 1]
```diff
@@ -72,7 +72,7 @@ harden(assertCapData);
  * @param {string} key
  * @param {ReturnType<typeof import('@endo/marshal').makeMarshal>['fromCapData']} fromCapData
  * @param {number} index index of the desired value in a deserialized stream cell
- * @return {any}
+ * @returns {any}
```
Comment (erights) [inline packages/internal/src/marshal.js]: This is a drive by, totally unrelated to this PR. I just happened to notice a lint warning.

File: packages/ERTP/src/typeGuards.js [hunk 2]
```diff
@@ -77,6 +77,20 @@ export const AmountShape = harden({
   value: AmountValueShape,
 });
 
+/**
+ * To be used as a guard for an amount pattern argument, i.e., an argument which
```
Comment (erights) [inline packages/ERTP/src/typeGuards.js]: "guard" terminology is wrong. It is a pattern, to be used in argument guard position of a method guard. TODO fix.

File: packages/vats/src/virtual-purse.js [hunk 3]
```diff
@@ -3,12 +3,13 @@ import { E } from '@endo/far';
...
 
 import {
+  AmountPatternShape,
   AmountShape,
   BrandShape,
   DepositFacetShape,
   NotifierShape,
   PaymentShape,
-} from '@agoric/ertp/src/typeGuards.js';
+} from '@agoric/ertp';
```
Comment (erights) [inline packages/vats/src/virtual-purse.js:14]: Truncating the import from `@agoric/ertp/src/typeGuards.js` to `@agoric/ertp` is mostly a drive-by. It is related to this PR only because I needed to add another import name.

**Review by dckc — COMMENTED** [review]
This fixes the bug, but until we have a plan to upgrade vat-bank, clients have to not depend on any fix. I don't know how to manage that if we merge to master at this time, so I defer to others to approve.

File: packages/vats/src/virtual-purse.js (hunk 3 above)
Comment (dckc) [reply packages/vats/src/virtual-purse.js:14]: looks like the vat-bank bundle already includes all of ERTP, so I guess this doesn't matter.

But I did a quick experiment with

```js
import { BrandShape } from '@agoric/ertp/src/typeGuards.js';
console.log(BrandShape);
```

vs importing the whole package:

```
1.0M bundles/bundle-ertp-fun-all.js
596K bundles/bundle-ertp-fun.js
```

`typeGuards.js` only depends on patterns (and hence pass-style, ...) but it doesn't depend on @agoric/store, notifier, nor zones.

Comment (erights) [reply packages/vats/src/virtual-purse.js:14]: Attn @kriskowal is there any way our bundler can do better tree shaking on its own without us needing to write more complex imports?

Comment (kriskowal) [reply packages/vats/src/virtual-purse.js:14]: Given our constraints, and even in general cases, I believe that writing and using narrowly-scoped modules is the best technique for keeping applications small. Automatic tree-shaking is too automatic. Anything better than what we have requires transforming sources.

For this reason, I discourage using tools like “lodash” or “underscore”. Even Q was a delicate balance where every additional method had to be very likely to be used by any application that used Q.

There is some wiggle room. The constraint is not that every import site must use every exported facility. The constraint is rather that a bundle should overall make use of all of the functionality it contains. So, “pass-style”, while large and indivisible, is okay to the extent that every application that uses “pass-style” is likely to benefit from most of it.

Though, I would like us to become more deliberate about the modules exported at the top-level of every package. Importing through `src` suggests (but does not indicate!) that we did not carefully consider these trade-offs in the design of the public API of the package.

**Review by turadg — CHANGES_REQUESTED** [review]
How does this fix #9407? I see only a refactor creating `AmountPatternShape` as an alias for `M.pattern()`. 

Please include a test showing the behavior that failed before this PR and passes after.

File: packages/vats/src/virtual-purse.js (hunk 3 above)
Comment (erights) [reply packages/vats/src/virtual-purse.js:14]: Since there's no issue here for this PR to address, resolving.

File: packages/vats/test/vpurse.test.js [hunk 4]
```diff
@@ -170,7 +171,7 @@ test('makeVirtualPurse', async t => {
 
   expected.pushAmount(fungible837);
   await E(vpurse)
-    .deposit(payment, fungible837)
+    .deposit(payment, M.and(fungible837))
```
Comment (erights) [inline packages/vats/test/vpurse.test.js:254]: > How does this fix #9407? I see only a refactor creating `AmountPatternShape` as an alias for `M.pattern()`.
> 
> Please include a test showing the behavior that failed before this PR and passes after.

Changing this to provide another pattern, that in this case happens to mean the same thing, causes this test to fail before this PR with

```
  Error {
    message: 'In "retain" method of (VirtualPurseKit utils): arg 1?: "[match:and]" - Must be a copyRecord to match a copyRecord pattern: {"brand":"[match:remotable]","value":"[match:or]"}',
  }

  › async Object.receive (packages/vats/src/virtual-purse.js:219:23)
```
which points to the line
```js
          const amt = await this.facets.utils.retain(payment, optAmountShape);
```
because prior to this PR the retain guard was declared
```js
    retain: M.callWhen(PaymentShape).optional(amountShape).returns(amountShape),
```
and the pattern is not itself an amountShape.

Done. PTAL.

File: packages/ERTP/src/typeGuards.js (hunk 2 above)
Comment (erights) [reply packages/ERTP/src/typeGuards.js]: Done

File: packages/internal/src/marshal.js (hunk 1 above)
Comment (erights) [reply packages/internal/src/marshal.js]: Resolving.

**Review by mhofman — CHANGES_REQUESTED** [review]
> ### Upgrade Considerations
> [#9402 (comment)](https://github.com/Agoric/agoric-sdk/pull/9402#discussion_r1613503324) and [#9402 (comment)](https://github.com/Agoric/agoric-sdk/pull/9402#discussion_r1613622732) explain the upgrade concern that was likely the cause for omitting this fix from #9042

The comment by @turadg states:
> My remaining concern is that if we make this fix and don't ship it for a while then in the meantime people may right code that works in dev but will fail when they get to the production chain (or test in a3p)

I'm not sure in what vat the `virtual-purse.js` which contains this guard change ends up in, but as long as it is upgraded before any of its consumers that assume the new behavior, we should be ok. Same story for `localchain`. Landing this change in master affectively requires us to upgrade these vats in the next release. As such, the next upgrade handler should have code to upgrade these vats explicitly (since we do not yet have generic code to upgrade all vats). This PR should introduce that change. Also since this change is somewhat subtle, I'm wondering if we need a code comment for future reviewers of release changes stating the nature of this specific change.

**Comment by erights** [comment]
Originally this PR was supposed to fix #9407 , by relaxing an incorrect rejection into a correct acceptance. But we are not yet ready to deploy such a fix on-chain, due to upgrade constraints, so fixing it in agoric-sdk master would be misleading. It might lead people to write other code that depends on this bug being fixed when tested against agoric-sdk master, only to have that code fail later on-chain when invoking the non-upgraded buggy code.

So we unfixed this PR to restore the buggy behavior, but with a better error message, and refactored to make fixing it again easier. #9863 is that fix. #9863  will remain draft until we are confident we can upgrade the relevant on-chain vat (the vbank) in a timely manner.

**Comment by erights** [comment]
@mhofman @turadg see https://github.com/Agoric/agoric-sdk/pull/9410#issuecomment-2274482994 for an explanation of a recent change. PTAL.

Attn @JD-Lorax

File: packages/vats/src/virtual-purse.js [hunk 5]
```diff
@@ -102,6 +105,22 @@ export const makeVirtualPurseKitIKit = (
...
+ * function restricts the `optAmountShape`, if provided, to be a concrete
+ * `Amount` rather than a `Pattern` as it is supposed to be.
+ *
+ * TODO: Once https://github.com/Agoric/agoric-sdk/issues/9407 is fixed, remove
+ * this function and all calls to it.
+ *
+ * @param {Pattern} [optAmountShape]
+ */
+const legacyRestrictAmountShapeArg = optAmountShape => {
+  if (optAmountShape && !matches(optAmountShape, AmountShape)) {
```
Comment (mhofman) [inline packages/vats/src/virtual-purse.js:119]: Have we verified that `makeVirtualPurseKitIKit` is never called with an `amountShape` different than `undefined` or `AmountShape` ?

File: packages/vats/src/virtual-purse.js [hunk 6]
```diff
@@ -48,15 +49,17 @@ export const makeVirtualPurseKitIKit = (
...
+      .returns(amountShape),
     redeem: M.callWhen(amountShape).returns(PaymentShape),
   });
 
   const UtilsI = M.interface('Utils', {
     retain: getInterfaceGuardPayload(RetainRedeemI).methodGuards.retain,
     redeem: getInterfaceGuardPayload(RetainRedeemI).methodGuards.redeem,
     recoverableClaim: M.callWhen(M.await(PaymentShape))
-      .optional(amountShape)
+      .optional(AmountPatternShape)
```
Comment (mhofman) [inline packages/vats/src/virtual-purse.js:62]: Should `AmountPatternShape` be a parameter to `makeVirtualPurseKitIKit` like the others?

File: packages/vats/test/vpurse.test.js [hunk 7]
```diff
@@ -176,6 +177,86 @@ test('makeVirtualPurse', async t => {
     .then(checkWithdrawal);
 });
 
+// TODO Once https://github.com/Agoric/agoric-sdk/issues/9407 is fixed,
+// This test should replace the similar one above.
+test.failing('makeVirtualPurse with optAmountShape pattern', async t => {
```
Comment (mhofman) [inline packages/vats/test/vpurse.test.js:182]: Is there anything else to do besides removing the `.failing` once fixed?

**Review by mhofman — COMMENTED** [review]
Reviewed purely on the upgradability aspect, and to match the approach we discussed. I didn't review if the choice of pattern itself is appropriate, nor the test itself.

Regarding the PR itself, from what I understanding this is now effectively a `refactor` and no longer a `fix`, right ?

File: packages/vats/src/virtual-purse.js (hunk 5 above)
Comment (erights) [reply packages/vats/src/virtual-purse.js:119]: I just looked. Yes. There are only two calls, and both have zero args, letting both params default.

Done.

File: packages/vats/src/virtual-purse.js (hunk 6 above)
Comment (erights) [reply packages/vats/src/virtual-purse.js:62]: No. It should always be exactly an alias for `M.pattern()` since any pattern should be valid.

Resolving.

File: packages/vats/test/vpurse.test.js (hunk 7 above)
Comment (erights) [reply packages/vats/test/vpurse.test.js:182]: The failing test is a trivial variant of the normal test above it, duplicated only to test this additional aspect. https://github.com/Agoric/agoric-sdk/pull/9863 deduplicates it, but has the one test keep the test of this additional aspect.

Resolving.

**Comment by erights** [comment]
@turadg , you requested changes at https://github.com/Agoric/agoric-sdk/pull/9410#pullrequestreview-2151726981 writing:

> How does this fix https://github.com/Agoric/agoric-sdk/issues/9407? I see only a refactor creating AmountPatternShape as an alias for M.pattern().
>
> Please include a test showing the behavior that failed before this PR and passes after.

That part is long done. If that remains your only concern and you're satisfied with the answer, could you please clear your requested changes so that I remain blocked only on @mhofman 's concerns? Thanks.

**Comment by erights** [comment]
> Regarding the PR itself, from what I understanding this is now effectively a `refactor` and no longer a `fix`, right ?

Good point. Done. PTAL.

**Review by turadg — APPROVED** [review]

**Review by mhofman — APPROVED** [review]
